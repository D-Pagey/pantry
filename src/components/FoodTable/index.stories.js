import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Fridge } from '../../fixtures';
import { FirebaseContext } from '../ProviderFirebase';
import { FoodTable } from '.';

const props = {
    food: Fridge,
    handleEdit: () => {},
    setFood: () => {}
};

const firebaseContext = {
    deleteFoodItem: () => {}
};

export default { title: 'FoodTable' };

export const normal = () => (
    <FirebaseContext.Provider value={{ ...firebaseContext }}>
        <MemoryRouter>
            <FoodTable {...props} />
        </MemoryRouter>
    </FirebaseContext.Provider>
);
