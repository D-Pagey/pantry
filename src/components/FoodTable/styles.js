import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';

export const Wrapper = styled.div`
    overflow: scroll;
    padding: 1rem 0;
`;

export const Table = styled.table`
    border: 1px solid #bdbdbd;
    border-collapse: collapse;
`;

export const TableHead = styled.thead`
    border-bottom: 1px solid grey;
`;

export const Item = styled.li`
    color: ${({ colour }) => colour};
    list-style: none;
`;

export const Header = styled.th`
    cursor: pointer;
    padding: 1rem;
`;

export const Link = styled(RouterLink)`
    cursor: pointer;
    color: blue;
    padding: 0 0 0 1rem;
`;

export const TableData = styled.td`
    padding: 1rem;
    background-color: #d7faff;
`;

export const TableRow = styled.tr`
    & > td {
        background-color: ${({ isOdd }) => isOdd && 'white'};
    }
`;
