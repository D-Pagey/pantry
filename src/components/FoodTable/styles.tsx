import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';

export const Wrapper = styled.div`
    overflow: scroll;
    padding: 1rem 0;
`;

export const Table = styled.table`
    border: 1px solid #bdbdbd;
    border-collapse: collapse;
`;

export const TableHead = styled.thead`
    border-bottom: 1px solid grey;
`;

type ItemProps = {
    readonly colour: string;
};

export const Item = styled.li<ItemProps>`
    color: ${({ colour }): string => colour};
    list-style: none;
`;

export const Header = styled.th`
    cursor: pointer;
    padding: 1rem;
`;

export const Link = styled(RouterLink)`
    cursor: pointer;
    color: blue;
    padding: 0 0 0 1rem;
`;

export const TableData = styled.td`
    padding: 1rem;
    background-color: #d7faff;
`;

type TableRowProps = {
    readonly isOdd: boolean;
};

export const TableRow = styled.tr<TableRowProps>`
    & > td {
        background-color: ${({ isOdd }) => isOdd && 'white'};
    }
`;
