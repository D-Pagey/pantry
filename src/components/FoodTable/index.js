import 'react-table/react-table.css';
import React, { useContext } from 'react';
import { shape, string } from 'prop-types';
import { differenceInDays, format } from 'date-fns';
import ReactTable from 'react-table';
import deleteIcon from '../../assets/delete.svg';
import { FirebaseContext } from '../ProviderFirebase';
import * as S from './styles';

const chooseColour = (date) => {
    const difference = differenceInDays(date, new Date());

    if (difference < 1) return 'red';
    if (difference <= 2) return 'blue';

    return 'black';
};

const FoodTable = ({ match }) => {
    const { category } = match.params;
    const { fridge, updateFridge } = useContext(FirebaseContext);

    const filteredData =
        category === 'all'
            ? fridge
            : fridge.filter((item) => {
                  return item.category === category;
              });

    const handleDelete = (name) => () => {
        const filteredItems = fridge.filter((item) => item.name !== name);
        updateFridge(filteredItems);
    };

    const columns = [
        {
            Header: 'Name',
            accessor: 'name'
        },
        {
            id: 'expires',
            Header: 'Expires',
            accessor: (item) => (
                <S.Item colour={chooseColour(item.expires)}>
                    {format(item.expires, 'do MMM')}
                </S.Item>
            )
        },
        {
            id: 'servings',
            Header: 'Servings',
            accessor: (item) => item.servings
        },
        {
            id: 'amend',
            Header: 'Amend',
            accessor: (item) => (
                <button
                    type="button"
                    onClick={handleDelete(item.name)}
                    style={{ cursor: 'pointer' }}
                    data-testid="deleteButton"
                >
                    <img src={deleteIcon} alt="delete" />
                </button>
            )
        }
    ];

    return (
        <div>
            <h1>{match.params.category}</h1>

            <ReactTable columns={columns} data={filteredData} defaultPageSize={10} />
        </div>
    );
};

FoodTable.propTypes = {
    match: shape({
        params: shape({
            category: string.isRequired
        }).isRequired
    }).isRequired
};

export default FoodTable;
