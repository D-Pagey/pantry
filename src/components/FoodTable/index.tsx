import React, { useContext, useEffect, useState } from 'react';
import { Link, useHistory, useParams, Redirect } from 'react-router-dom';
import { format } from 'date-fns';
import arraySort from 'array-sort';

import { chooseDateColour, doesCategoryExist } from '../../utils';
import deleteIcon from '../../assets/delete.svg';
import editIcon from '../../assets/edit.svg';
import { FirebaseContext } from '../ProviderFirebase';
import Button from '../Button';
import * as S from './styles';

type itemTypes = {
    category: { label: string; colour: string };
    expires: Date;
    id: string;
    name: string;
    servings: number;
};

const FoodTable = (): JSX.Element => {
    const [isValidCategory, setIsValidCategory] = useState<boolean>();
    const [data, setData] = useState([]);
    const history = useHistory();
    const [isDescendingOrder, setIsDescendingOrder] = useState(false);
    const { category } = useParams();
    const { categories, fridge, updateHousehold } = useContext(FirebaseContext);

    useEffect(() => {
        if (categories.length > 0 && category) {
            setIsValidCategory(doesCategoryExist(categories, category));
        }
    }, [categories, category]);

    useEffect(() => {
        if (data.length === 0) {
            const filteredData =
                category === 'all'
                    ? fridge
                    : fridge.filter((item: { category: { label: string; colour: string } }) => {
                          return item.category.label === category;
                      });

            setData(filteredData);
        }
    }, [data.length, category, fridge]);

    const handleDelete = (id: string) => (): void => {
        const filteredItems = fridge.filter((item: { id: string }) => item.id !== id);
        updateHousehold({ key: 'fridge', values: filteredItems, isDeleting: true });
    };

    const handleEdit = (params: itemTypes) => (): void => {
        history.push('/add', params);
    };

    const expiresColumn = (item: itemTypes): JSX.Element => (
        <S.Item colour={chooseDateColour(item.expires)}>{format(item.expires, 'do MMM')}</S.Item>
    );

    const sortData = (property: string) => (): void => {
        const sorted = arraySort([...data], property, { reverse: isDescendingOrder });

        setIsDescendingOrder(!isDescendingOrder);
        setData(sorted);
    };

    if (isValidCategory === false) return <Redirect to="/not-found" />;

    return (
        <div>
            <h1>{category}</h1>

            <table>
                <thead>
                    <tr>
                        <S.Header onClick={sortData('name')}>Name</S.Header>
                        <S.Header onClick={sortData('expires')}>Expires</S.Header>
                        {category === 'all' && (
                            <S.Header
                                data-testid="foodTableCategoryColumn"
                                onClick={sortData('category.label')}
                            >
                                Category
                            </S.Header>
                        )}
                        <S.Header onClick={sortData('servings')}>Servings</S.Header>
                        <S.Header>Amend</S.Header>
                    </tr>
                </thead>
                <tbody>
                    {data.map((item: itemTypes) => {
                        return (
                            <tr key={item.id}>
                                <td>{item.name}</td>
                                <td>{expiresColumn(item)}</td>
                                {category === 'all' && (
                                    <td>
                                        <Link to={`/food/${item.category.label}`}>
                                            {item.category.label}
                                        </Link>
                                    </td>
                                )}
                                <td>{item.servings}</td>
                                <td>
                                    <button
                                        type="button"
                                        onClick={handleEdit(item)}
                                        style={{ cursor: 'pointer', margin: '0 1rem 0 0' }}
                                        data-testid="editButton"
                                    >
                                        <img src={editIcon} alt="edit" />
                                    </button>
                                    <button
                                        type="button"
                                        onClick={handleDelete(item.id)}
                                        style={{ cursor: 'pointer' }}
                                        data-testid="deleteButton"
                                    >
                                        <img src={deleteIcon} alt="delete" />
                                    </button>
                                </td>
                            </tr>
                        );
                    })}
                </tbody>
            </table>

            <Link to="/add">
                <Button>Add Item</Button>
            </Link>
        </div>
    );
};

export default FoodTable;
