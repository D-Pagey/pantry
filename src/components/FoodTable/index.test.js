import React from 'react';
import userEvent from '@testing-library/user-event';
import { addDays, format } from 'date-fns';
import FoodTable from '.';

const props = {
    match: {
        params: {
            category: 'meat'
        }
    }
};

const firebaseContext = {
    fridge: [
        {
            category: 'meat',
            expires: new Date(2019, 9, 12),
            name: 'chicken',
            servings: 2
        },
        {
            category: 'fish',
            expires: new Date(2019, 3, 9),
            name: 'salmon',
            servings: 1
        }
    ],
    updateFridge: () => {}
};

describe('FoodTable component', () => {
    it('should render', () => {
        const { container } = render(<FoodTable {...props} />, firebaseContext);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should handle delete', async () => {
        const updateFridge = jest.fn();
        const { queryAllByTestId } = render(<FoodTable {...props} />, {
            ...firebaseContext,
            updateFridge
        });
        const deleteButton = queryAllByTestId('deleteButton');

        userEvent.click(deleteButton[0]);

        expect(updateFridge).toHaveBeenCalledWith(
            firebaseContext.fridge.filter((item, index) => index !== 0)
        );
    });

    it.each`
        colour     | date
        ${'red'}   | ${new Date()}
        ${'blue'}  | ${addDays(new Date(), 2)}
        ${'black'} | ${addDays(new Date(), 4)}
    `('should have $colour for expiry date', ({ colour, date }) => {
        const item = {
            category: 'meat',
            expires: date,
            name: 'chicken',
            servings: 2
        };
        const updatedContext = {
            ...firebaseContext,
            fridge: [item]
        };

        const { getByText } = render(<FoodTable {...props} />, updatedContext);
        const expiryDate = getByText(format(date, 'do MMM'));

        expect(expiryDate).toHaveStyleRule('color', colour);
    });

    it('should handle the category: all', () => {
        const updatedProps = {
            ...props,
            match: {
                params: {
                    category: 'all'
                }
            }
        };
        const { getByText } = render(<FoodTable {...updatedProps} />, firebaseContext);

        getByText(firebaseContext.fridge[0].name);
        getByText(firebaseContext.fridge[1].name);
    });
});
