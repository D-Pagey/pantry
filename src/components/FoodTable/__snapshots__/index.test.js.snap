// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodTable component should render 1`] = `
.c0 {
  overflow: scroll;
  padding: 1rem 0;
}

.c1 {
  border: 1px solid #bdbdbd;
  border-collapse: collapse;
}

.c2 {
  border-bottom: 1px solid grey;
}

.c6 {
  color: red;
  list-style: none;
}

.c3 {
  cursor: pointer;
  padding: 1rem;
}

.c5 {
  cursor: pointer;
  color: blue;
  padding: 0 0 0 1rem;
}

.c4 {
  padding: 1rem;
  background-color: #d7faff;
}

.c7 > td {
  background-color: white;
}

<div
  class="c0"
>
  <table
    class="c1"
  >
    <thead
      class="c2"
    >
      <tr>
        <th
          class="c3"
        >
          Name
        </th>
        <th
          class="c3"
          data-testid="foodTableCategoryColumn"
        >
          Categories
        </th>
        <th
          class="c3"
        >
          Expires
        </th>
        <th
          class="c3"
        >
          Servings
        </th>
        <th
          class="c3"
        >
          Amend
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
      >
        <td
          class="c4"
          data-testid="foodTableName"
        >
          chicken
        </td>
        <td
          class="c4"
        >
          <a
            class="c5"
            href="/food/111"
          >
            111
          </a>
        </td>
        <td
          class="c4"
        >
          <li
            class="c6"
          >
            12th Oct
          </li>
        </td>
        <td
          class="c4"
        >
          2
        </td>
        <td
          class="c4"
        >
          <button
            data-testid="editButton"
            style="cursor: pointer; margin: 0px 1rem 0px 0px;"
            type="button"
          >
            <img
              alt="edit"
              src="edit.svg"
            />
          </button>
          <button
            data-testid="deleteButton0"
            style="cursor: pointer;"
            type="button"
          >
            <img
              alt="delete"
              src="delete.svg"
            />
          </button>
        </td>
      </tr>
      <tr
        class="c7"
      >
        <td
          class="c4"
          data-testid="foodTableName"
        >
          steak
        </td>
        <td
          class="c4"
        >
          <a
            class="c5"
            href="/food/111"
          >
            111
          </a>
        </td>
        <td
          class="c4"
        >
          <li
            class="c6"
          >
            9th Apr
          </li>
        </td>
        <td
          class="c4"
        >
          1
        </td>
        <td
          class="c4"
        >
          <button
            data-testid="editButton"
            style="cursor: pointer; margin: 0px 1rem 0px 0px;"
            type="button"
          >
            <img
              alt="edit"
              src="edit.svg"
            />
          </button>
          <button
            data-testid="deleteButton1"
            style="cursor: pointer;"
            type="button"
          >
            <img
              alt="delete"
              src="delete.svg"
            />
          </button>
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="c4"
          data-testid="foodTableName"
        >
          salmon
        </td>
        <td
          class="c4"
        >
          <a
            class="c5"
            href="/food/222"
          >
            222
          </a>
        </td>
        <td
          class="c4"
        >
          <li
            class="c6"
          >
            9th Apr
          </li>
        </td>
        <td
          class="c4"
        >
          1
        </td>
        <td
          class="c4"
        >
          <button
            data-testid="editButton"
            style="cursor: pointer; margin: 0px 1rem 0px 0px;"
            type="button"
          >
            <img
              alt="edit"
              src="edit.svg"
            />
          </button>
          <button
            data-testid="deleteButton2"
            style="cursor: pointer;"
            type="button"
          >
            <img
              alt="delete"
              src="delete.svg"
            />
          </button>
        </td>
      </tr>
      <tr
        class="c7"
      >
        <td
          class="c4"
          data-testid="foodTableName"
        >
          broccolli
        </td>
        <td
          class="c4"
        >
          <a
            class="c5"
            href="/food/111"
          >
            111
          </a>
          <a
            class="c5"
            href="/food/333"
          >
            333
          </a>
        </td>
        <td
          class="c4"
        >
          <li
            class="c6"
          >
            9th Apr
          </li>
        </td>
        <td
          class="c4"
        >
          1
        </td>
        <td
          class="c4"
        >
          <button
            data-testid="editButton"
            style="cursor: pointer; margin: 0px 1rem 0px 0px;"
            type="button"
          >
            <img
              alt="edit"
              src="edit.svg"
            />
          </button>
          <button
            data-testid="deleteButton3"
            style="cursor: pointer;"
            type="button"
          >
            <img
              alt="delete"
              src="delete.svg"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
